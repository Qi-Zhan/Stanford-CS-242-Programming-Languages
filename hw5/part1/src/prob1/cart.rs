/***************
 * NOTE:
 *   You can call `internal_login` simply by `internal_login(id, pw)`.
 ***************/

#[allow(unused_imports)]
use crate::prob1::server::internal_login;

// TODO: Implement the following typestate structs. You are free to add any fields to the struct
// definitions.
// Below shows the signature of the methods to be implemented across the cart state machine.
// Note that not all methods may be implemented for every typestate struct.
//
//   pub fn login(_: String, _: String) -> Result<T,()>
//   pub fn add_item(self, _: u32) -> T
//   pub fn clear_items(self) -> T
//   pub fn checkout(self) -> T
//   pub fn cancel(self) -> T
//   pub fn order(self) -> T
//   pub fn acct_num(&self) -> u32
//   pub fn tot_cost(&self) -> u32
//
// Here T denotes a type. Note that each T can be a different type.
//===== BEGIN_CODE =====//
pub struct Cart {  }
pub struct Empty {
    acct_num_val : u32 ,
    tot_cost_val : u32
}
pub struct NonEmpty {
    acct_num_val : u32 ,
    tot_cost_val : u32
}
pub struct Checkout {
    acct_num_val : u32 ,
    tot_cost_val : u32
}

impl Cart {
    pub fn login(id: String, pw: String) -> Result<Empty,()> {
        let user: Option<u32> = internal_login(id, pw);
        match user {
            None => return Err(()),
            Some(val) => return Ok(Empty{acct_num_val: val, tot_cost_val: 0}),
        }
    }
}

impl Empty {
    pub fn add_item(self, cost: u32) -> NonEmpty {
        return NonEmpty{
            acct_num_val: self.acct_num_val,
            tot_cost_val: self.tot_cost_val + cost
        };
    }

    pub fn acct_num(&self) -> u32 {
        return self.acct_num_val
    }

    pub fn tot_cost(&self) -> u32 {
        return self.tot_cost_val
    }
}

impl NonEmpty {

    pub fn add_item(self, cost: u32) -> NonEmpty {
        return NonEmpty{
            acct_num_val: self.acct_num_val,
            tot_cost_val: self.tot_cost_val + cost
        };
    }

    pub fn checkout(self) -> Checkout {
        return Checkout{
            acct_num_val: self.acct_num_val,
            tot_cost_val: self.tot_cost_val
        };
    }

    pub fn clear_items(self) -> Empty {
        return Empty {
            acct_num_val: self.acct_num_val ,
            tot_cost_val: 0
        };
    }

    pub fn acct_num(&self) -> u32 {
        return self.acct_num_val;
    }

    pub fn tot_cost(&self) -> u32 {
        return self.tot_cost_val;
    }
}

impl Checkout {

    pub fn order(self) -> Empty {
        return Empty {
            acct_num_val : self.acct_num_val,
            tot_cost_val : 0,
        }
    }

    pub fn cancel(self) -> NonEmpty {
        return NonEmpty {
            acct_num_val : self.acct_num_val,
            tot_cost_val : self.tot_cost_val,
        }
    }

    pub fn acct_num(&self) -> u32 {
        return self.acct_num_val
    }

    pub fn tot_cost(&self) -> u32 {
        return self.tot_cost_val
    }

}


//===== END_CODE =====//
