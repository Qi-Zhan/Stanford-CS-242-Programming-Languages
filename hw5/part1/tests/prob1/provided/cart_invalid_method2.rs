extern crate part1;
#[allow(unused_imports)]
use part1::prob1::cart::Cart;

fn main() {
    match Cart::login("id1".to_string(), "pw1".to_string()) {
        Err(()) =>
            assert!(false),
        Ok(empty) => {
            let nonempty = empty.add_item(1);
            let checkout = nonempty.checkout();
            let nonempty = checkout.cancel();
            // `order` method should be inaccessible in nonempty state.
            nonempty.order();
        }
    }
}
