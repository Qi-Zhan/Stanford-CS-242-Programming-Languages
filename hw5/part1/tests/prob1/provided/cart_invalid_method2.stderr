error[E0599]: no method named `order` found for struct `NonEmpty` in the current scope
  --> tests/prob1/provided/cart_invalid_method2.rs:14:22
   |
14 |             nonempty.order();
   |                      ^^^^^ method not found in `NonEmpty`
